### PR TITLE
Ranking - selectToRank PR

### PR DESCRIPTION
**Pull request by dmitrykurmanov**
see #5328

### CODE REVIEW

File: src/question_ranking.ts [hunk 1]
```diff
@@ -64,6 +67,27 @@ export class QuestionRankingModel extends QuestionCheckboxModel {
       .toString();
   }
 
+  public getContainerClasses(containerType?: string) {
+    let isEmpty = false;
```
Comment (andrewtelnov) [inline src/question_ranking.ts:71]:
```
    const isToContainer = containerType === "to";
    const isFromContainer  = containerType === "from";
    let isEmpty = false;

    if (isToContainer) {
      isEmpty = this.rankingChoices.length === 0;
    } else if (isFromContainer ) {
      isEmpty = this.unRankingChoices.length === 0;
    }
```

File: src/dragdrop/ranking-select-to-rank.ts [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    const questionModel: any = this.parentElement;
+    const rankingChoices = questionModel.rankingChoices;
+    const unRankingChoices = questionModel.unRankingChoices;
+
+    let fromIndex;
+    let toIndex;
+
+    if (this.isDraggedElementUnranked && this.isDropTargetRanked) {
+      fromIndex = unRankingChoices.indexOf(this.draggedElement);
+      if (rankingChoices.length === 0) {
```
Comment (andrewtelnov) [inline src/dragdrop/ranking-select-to-rank.ts:69]:
```
toIndex = rankingChoices.indexOf(this.dropTarget);
if(toIndex < 0) toIndex = 0;

File: src/dragdrop/ranking-select-to-rank.ts [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        toIndex = 0;
+      } else {
+        toIndex = rankingChoices.indexOf(this.dropTarget);
+      }
+      this.selectToRankEnabled(questionModel, fromIndex, toIndex);
+      this.doUIEffects(dropTargetNode, fromIndex, toIndex);
+      return;
+    }
+
+    if (this.isDraggedElementRanked && this.isDropTargetRanked) {
```
Comment (andrewtelnov) [inline src/dragdrop/ranking-select-to-rank.ts:79]: Code duplication, create a function that will pass choices (ranking or unranking) and function one of the function: selectToRankEnabled or reorderRankedItem or unselectFromRank.

File: src/dragdrop/ranking-select-to-rank.ts [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    return !this.isDropTargetRanked;
+  }
+
+  public selectToRankEnabled(questionModel: QuestionRankingModel, fromIndex: number, toIndex: number): void {
+    const rankingChoices = questionModel.rankingChoices;
+    const unRankingChoices = questionModel.unRankingChoices;
+    const item = unRankingChoices[fromIndex];
+
+    rankingChoices.splice(toIndex, 0, item);
+    questionModel.setPropertyValue("rankingChoices", rankingChoices);
```
Comment (andrewtelnov) [inline src/dragdrop/ranking-select-to-rank.ts:146]: After fix the code duplication the last line and probably some others can be moved to another function and do not used in all 3 functions.

File: src/question_ranking.ts [hunk 5]
```diff
@@ -269,11 +354,80 @@ export class QuestionRankingModel extends QuestionCheckboxModel {
...
+    const rankingChoices = this.rankingChoices;
+    const unRankingChoices = this.unRankingChoices;
+
+    const isMovedElementRanked = rankingChoices.indexOf(movedElement) !== -1;
+    const isMovedElementUnRanked = !isMovedElementRanked;
+
+    let fromIndex;
+    let toIndex;
+
+    if (key === " " && isMovedElementUnRanked) {
```
Comment (andrewtelnov) [inline src/question_ranking.ts:371]: I feel that some code duplication here can be removed as well.

File: visualRegressionTests/tests/defaultV2/etalons/question-ranking-select-to-rank-vertical_mask.png
Comment (andrewtelnov) [inline visualRegressionTests/tests/defaultV2/etalons/question-ranking-select-to-rank-vertical_mask.png:1]: We don't need mask files

File: src/question_ranking.ts [hunk 6]
```diff
@@ -177,13 +236,34 @@ export class QuestionRankingModel extends QuestionCheckboxModel {
...
+
   public dragDropRankingChoices: DragDropRankingChoices;
   @property({ defaultValue: null }) currentDropTarget: ItemValue;
   @property({ defaultValue: null }) dropTargetNodeMove: string;
 
   endLoadingFromJson(): void {
     super.endLoadingFromJson();
-    this.dragDropRankingChoices = new DragDropRankingChoices(this.survey, null, this.longTap);
+
+    if (this.selectToRankEnabled) {
```
[the remark this answers is not in the document]
Comment (dmitrykurmanov) [reply src/question_ranking.ts]: the fix: https://github.com/surveyjs/survey-library/commit/1a23176cb6a713052a50660a0702b5d2ada9f66c

**Pull request merged**